packages/synapse-sdk/src/telemetry/service.ts

```diff
@@ -0,0 +1,443 @@
+  ua?: string
+  appName?: string
+}
+
+/**
+ * Main telemetry service that manages the adapter and provides high-level APIs
+ */
+export class TelemetryService {
+  private adapter: TelemetryAdapter
+  private enabled: boolean
+  private context: RuntimeContext
+  private eventBuffer: Array<ErrorEvent | HTTPEvent | OperationEvent | CustomEvent> = []
+  private readonly maxBufferSize = 50
+
+  constructor(adapter: TelemetryAdapter, config: TelemetryConfig, context: RuntimeContext) {
+    this.adapter = adapter
+
+    // Resolve configuration with environment detection
+    const resolvedConfig = resolveTelemetryConfig(config)
+    this.enabled = resolvedConfig.enabled
+    this.context = context
+
+    if (this.enabled) {
+      this.adapter.init(resolvedConfig, {
+        sdkVersion: context.sdkVersion,
+        runtime: context.runtime,
+        network: context.network,
+        ua: context.ua || '',
+        appName: context.appName || '',
+      })
+
+      this.setupIntervalFlushing()
+    }
+  }
+
+  /**
+   * We will automatically flush every 5 seconds so we don't miss events if a process calls `process.exit(code)`
+   * without calling `await synapse.telemetry.flush()`.
+   *
+   * Uses setTimeout instead of setInterval to avoid keeping the process alive.
+   */
+  private setupIntervalFlushing(): void {
+    const scheduleNext = () => {
+      setTimeout(() => {
+        this.flush().finally(() => {
+          // Schedule the next flush only after the current one completes
+          scheduleNext()
+        })
+      }, 5000)
+    }
+
+    // Start the first flush cycle
+    scheduleNext()
+  }
+
+  /**
+   * Track an entire operation with timing and error handling
+   *
+   * This is the primary way to instrument SDK methods. Wrap your method body:
+   *
+   * @example
+   * ```typescript
+   * async myMethod() {
+   *   return this.telemetry.trackOperation('service.myMethod', async () => {
+   *     const result = await doWork()
+   *     return result
+   *   })
+   * }
+   * ```
+   *
+   * @param operation - Operation type (use OPERATIONS constants)
+   * @param fn - Async function to execute and track
+   * @param params - Optional allowlisted parameters (no secrets!)
+   * @returns Result from the function
+   */
+  async trackOperation<T>(operation: OperationType, fn: () => Promise<T>, params?: Record<string, any>): Promise<T> {
+    if (!this.enabled) {
+      return fn()
+    }
+
+    const startTime = Date.now()
+    const requestId = this.generateRequestId()
+
+    try {
+      const result = await fn()
+
+      const event: OperationEvent = {
+        type: 'operation',
+        operation,
+        params: params ? this.sanitizeParams(params) : {},
+        success: true,
+        durationMs: Date.now() - startTime,
+        requestId,
+        ts: new Date().toISOString(),
+      }
+
+      this.addToBuffer(event)
+      this.adapter.captureOperation(event)
+
+      return result
+    } catch (error) {
```

Comment:
how much do we even need this?  Are any errors even expected?  Do we just let our global error catcher handle it?